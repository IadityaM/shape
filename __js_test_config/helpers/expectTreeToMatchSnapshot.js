import toJson from 'enzyme-to-json'

const expectTreeToMatchSnapshot = tree => {
  expect(toJson(tree)).toMatchSnapshot()
}

export default expectTreeToMatchSnapshot
