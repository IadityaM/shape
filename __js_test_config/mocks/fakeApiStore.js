const fakeApiStore = ({ findResult = '', findAllResult = [], requestResult = '' } = {}) => {
  return {
    add: jest.fn().mockName('add'),
    removeAll: jest.fn().mockName('removeAll'),
    find: jest.fn()
      .mockReturnValue(findResult),
    findAll: jest.fn()
      .mockReturnValue(findAllResult),
    request: jest.fn()
      .mockReturnValue(Promise.resolve(requestResult))
      .mockName('request'),
    fetch: jest.fn()
      .mockReturnValue(Promise.resolve())
      .mockName('fetch'),
    sync: jest.fn().mockName('sync'),
    setCurrentUserId: jest.fn(),
    currentUserId: 1,
    currentUser: {
      id: 1,
      current_user_collection_id: 99,
      current_organization: {
        name: 'test org'
      },
      name: 'Johnny Appleseed',
      pic_url_square: 'https://s3.amazonaws.com/pic.png',
      groups: [],
      terms_accepted: true,
    },
    collections: [],
    items: [],
  }
}

export default fakeApiStore
