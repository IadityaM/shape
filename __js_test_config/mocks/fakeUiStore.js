import { observable } from "mobx"

const fakeUiStore = {
  gridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  defaultGridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  blankContentToolState: {
    order: null,
    width: null,
    height: null,
    replacingId: null,
    placeholderCard: null,
  },
  dialogConfig: {
    open: null,
    prompt: null,
    onConfirm: null,
    onCancel: null,
    iconName: null,
    confirmText: 'OK',
    cancelText: 'Cancel',
    onClose: jest.fn(),
  },
  activityLogOpen: false,
  activityLogPosition: {
    x: 0,
    y: 0,
    h: 1,
    w: 1,
  },
  activityLogMoving: false,
  cardMenuOpen: {
    id: false,
    x: 0,
    y: 0,
    offsetX: 0,
    offsetY: 0,
  },
  selectedArea: { minX: null, maxX: null, minY: null, maxY: null },
  openBlankContentTool: jest.fn(),
  closeBlankContentTool: jest.fn(),
  setBctPlaceholderCard: jest.fn(),
  closeCardMenu: jest.fn(),
  openCardMenu: jest.fn(),
  closeMoveMenu: jest.fn(),
  resetSelectionAndBCT: jest.fn(),
  resetCardPositions: jest.fn(),
  rolesMenuOpen: false,
  isLoading: false,
  dismissedMoveHelper: false,
  selectedAreaEnabled: false,
  setSelectedArea: jest.fn(),
  selectedCardIds: [],
  selectCardId: jest.fn(),
  deselectCards: jest.fn(),
  setViewingRecord: jest.fn(),
  viewingCollection: null,
  isViewingHomepage: false,
  viewingItem: null,
  movingFromCollectionId: null,
  movingCardIds: [],
  openMoveMenu: jest.fn(),
  update: jest.fn(),
  alert: jest.fn(),
  alertOk: jest.fn(),
  defaultAlertError: jest.fn(),
  confirm: jest.fn(),
  closeDialog: jest.fn(),
  cardAction: 'move',
  blurContent: false,
  organizationMenuPage: 'organizationMenuPage',
  organizationMenuGroupId: null,
  organizationMenuOpen: false,
  expandedThreadKey: null,
  zoomLevel: 1,
  expandThread: jest.fn(),
  openGroup: jest.fn(),
  openOptionalMenus: jest.fn(),
  trackEvent: jest.fn(),
  trackedRecords: {},
  editingName: observable([]),
  activityLogPage: 'comments',
  pageMenuOpen: false,
  searchText: '',
  collectionCardSortOrder: '',
  addNewCard: jest.fn(),
  removeNewCard: jest.fn(),
  isNewCard: jest.fn(),
  autocompleteMenuClosed: jest.fn(),
  captureKeyboardGridClick: jest.fn(),
  popupAlert: jest.fn(),
  stopDragging: jest.fn(),
  multiMoveCardIds: [],
  setSnoozeChecked: jest.fn(),
  scrollToTop: jest.fn(),
  scrollToBottom: jest.fn(),
  scrollToBottomOfComments: jest.fn(),
  scrollToBottomOfModal: jest.fn(),
  popupSnackbar: jest.fn(),
  showPermissionsAlert: jest.fn(),
  gridHeightFor: jest.fn().mockReturnValue(250),
  performActionAfterRoute: jest.fn(),
  linkedBreadcrumbTrailForRecord: jest
    .fn()
    .mockImplementation(x => x.breadcrumb),
  addEmptySpaceClickHandler: jest.fn(),
  removeEmptySpaceClickHandler: jest.fn(),
  setEditingCardCover: jest.fn(),
  adminAudienceMenuOpen: false,
  actionMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  textMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  replyingToCommentId: false,
  setReplyingToComment: jest.fn(),
  setCommentingOnRecord: jest.fn(),
  setActivityLogPage: jest.fn(),
  setBodyBackgroundImage: jest.fn(),
  setBodyFontColor: jest.fn(),
  selectedTextRangeForCard: jest.fn(),
  reselectCardIds: jest.fn(),
  reselectOnlyEditableRecords: jest.fn(),
  reselectOnlyMovableCards: jest.fn(),
  setMovingCards: jest.fn(),
  setVisibleCols: jest.fn(),
  setVisibleRows: jest.fn(),
  setPlaceholderSpot: jest.fn(),
  setDroppingFilesCount: jest.fn(),
  drag: jest.fn(),
  startDragging: jest.fn(),
  adjustZoomLevel: jest.fn(),
  zoomIn: jest.fn(),
  zoomOut: jest.fn(),
  determineZoomLevels: jest.fn(),
  maxCols: jest.fn(),
  maxGridWidth: jest.fn().mockReturnValue(1384),
  pageMargins: jest.fn().mockReturnValue({ left: 100, top: 50 }),
  relativeZoomLevel: 2,
  zoomLevels: [],
  isSelected: jest.fn(),
  toggleSelectedCardId: jest.fn(),
  openContextMenu: jest.fn(),
  clearTextEditingItem: jest.fn(),
  positionForCoordinates: jest.fn().mockReturnValue({ x:0, y:0, xPos: 0, yPos: 0, width: 1, height: 1 }),
  dragGridSpot: new Map(),
  droppingFilesCount: 0,
}

export default fakeUiStore
