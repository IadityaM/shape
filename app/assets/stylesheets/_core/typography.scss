html {
  // we were just using bourbon for font-smoothing, so have included explicitly here
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 16px;
  font-family: 'Sentinel', serif;
}

@media only screen and (max-width: 645px) {
  html {
    font-size: 14px;
  }
}

@font-face {
  font-family: 'Sentinel';
  src: asset-url('Sentinel-Medium.otf');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: 'Sentinel';
  src: asset-url('Sentinel-MediumItal.otf');
  font-weight: normal;
  font-style: italic;
}

@font-face {
  font-family: 'Sentinel';
  src: asset-url('Sentinel-Book.otf');
  font-weight: 400;
  font-style: normal;
}

@font-face {
  font-family: 'Gotham';
  src: asset-url('Gotham-Medium.otf');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: 'Gotham';
  src: asset-url('Gotham-Book.otf');
  font-weight: 100;
  font-style: normal;
}
