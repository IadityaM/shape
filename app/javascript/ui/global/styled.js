import styled from 'styled-components'
import v from '~/utils/variables'

export const BctButton = styled.button`
`
BctButton.displayName = 'BctButton'

// Form Styles

export const FormActionsContainer = styled.div`
  padding-bottom: 14px;
  text-align: center;
`
FormActionsContainer.displayName = 'StyledFormActionsContainer'

export const FieldContainer = styled.div`
  padding-bottom: 35px;

  label {
    margin-right: 15px;
  }
`
FieldContainer.displayName = 'StyledFieldContainer'

export const Label = styled.label`
  text-transform: uppercase;
  margin-bottom: 20px;
  font-family: ${v.fonts.sans};
  font-size: 1.25rem;
  font-weight: ${v.weights.medium};
  letter-spacing: 1px;
  display: block;
`
Label.displayName = 'StyledLabel'

export const FormButton = styled.button`
  width: 183px;
  text-transform: uppercase;
  font-family: ${v.fonts.sans};
  font-size: 1rem;
  font-weight: 500;
  letter-spacing: 0.09375rem;
  height: 40px;
  cursor: pointer;
  color: #ffffff;
  border-radius: 20px;
  border: none;
  background-color: ${v.colors.blackLava};
`
FormButton.displayName = 'StyledFormButton'

export const TextField = styled.input`
  width: 225px;
  padding-right: 4px;
  padding-left: 4px;
  padding-bottom: 6px;
  outline-width: 0;
  font-size: 1rem;
  border: 0;
  border-bottom: 0.5px solid ${v.colors.gray};

  &::placeholder {
    color: ${v.colors.gray};
  }

  &:focus {
    outline-width: 0;
  }
`
TextField.displayName = 'StyledTextField'

const ImageField = styled.span`
  width: 100px;
  position: relative;
  height: 100px;
  display: block;
  color: white;
  border-radius: 50%;
  background-color: ${v.colors.gray};

  span {
    position: absolute;
    font-size: 36px;
    font-weight: ${v.weights.book};
    left: calc(50% - 8px);
    top: calc(50% - 23px);
  }
`
ImageField.displayName = 'StyledImageField'
