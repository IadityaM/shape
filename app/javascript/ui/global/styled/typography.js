import styled from 'styled-components'
import v from '~/utils/variables'

/** @component */
export const Heading1 = styled.h1`
  font-family: ${v.fonts.sans};
  font-size: 2.25rem;
  font-weight: ${v.weights.medium};
  letter-spacing: 2px;
  color: black;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
  white-space: ${props => (props.wrapLine ? 'normal' : 'nowrap')}; /* better this way for responsive? */
  text-transform: uppercase;

  @media only screen and (max-width: ${v.responsive.smallBreakpoint}px) {
    padding: 1rem 0;
    font-size: 1.5rem;
  }
`
Heading1.displayName = 'Heading1'

/** @component */
export const SimpleHeading1 = Heading1.extend`
  margin-bottom: 2rem;
  text-transform: none;
`
SimpleHeading1.displayName = 'SimpleHeading1'

/** @component */
export const Heading2 = styled.h2`
  text-transform: uppercase;
  font-family: ${v.fonts.sans};
  font-size: 1.5rem;
  font-weight: ${v.weights.medium};
  letter-spacing: 0.14375rem;
  color: ${v.colors.blackLava};
`
Heading2.displayName = 'Heading2'

/** @component */
export const Heading3 = styled.h3`
  text-transform: uppercase;
  margin-bottom: 13px;
  font-size: 0.9375rem;
  font-weight: ${v.weights.medium};
  letter-spacing: 0.0625rem;
`
Heading3.displayName = 'StyledHeading3'

/** @component */
export const DisplayText = styled.span`
  font-weight: ${v.weights.book};
  font-family: ${v.fonts.sans};
  font-size: 1rem;
`
DisplayText.displayName = 'StyledDisplayText'

/** @component */
export const SubduedText = DisplayText.extend`
  color: ${v.colors.cloudy};
`
SubduedText.displayName = 'SubduedText'

/** @component */
export const DisplayLink = styled.a`
  color: ${v.colors.gray};
  font-weight: ${v.weights.book};
  font-family: ${v.fonts.sans};
  font-size: 1rem;
  color: black;
`
DisplayLink.displayName = 'StyledDisplayLink'

/** @component */
export const SubduedTitle = styled.span`
  color: ${v.colors.gray};
  font-weight: ${v.weights.book};
  font-family: ${v.fonts.sans};
  font-size: 1rem;
`
SubduedTitle.displayName = 'StyledSubduedTitled'

/** @component */
export const ConfirmText = styled.span`
  font-weight: ${v.weights.book};
  font-family: ${v.fonts.sans};
  font-size: 1.25rem;
`

/** @component */
export const SubText = styled.span`
  vertical-align: super;
  font-family: ${v.fonts.serif};
  font-size: 0.75rem;
`
SubText.displayName = 'StyledSubText'

/** @component */
export const Anchor = styled.a`
  color: ${v.colors.blue};
  text-decoration: none;
  &:hover {
    text-decoration: underline;
  }
`
Anchor.displayName = 'StyledAnchor'
