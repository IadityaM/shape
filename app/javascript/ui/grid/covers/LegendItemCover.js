import _ from 'lodash'
import { action, observable } from 'mobx'
import { inject, observer, PropTypes as MobxPropTypes } from 'mobx-react'
import color from 'color'
import styled from 'styled-components'

import { Select, SelectOption } from '~/ui/global/styled/forms'
import AutoComplete from '~/ui/global/AutoComplete'
import XIcon from '~/ui/icons/XIcon'
import v from '~/utils/variables'
import { colorScale } from '~/ui/global/charts/ChartUtils'
import {
  DisplayText,
  Heading3,
  SmallHelperText,
} from '~/ui/global/styled/typography'
import LineChartIcon from '~/ui/icons/LineChartIcon'
import trackError from '~/utils/trackError'

function formatForAutocomplete(objects) {
  return objects.map(object => ({
    value: object.id,
    label: object.name,
    data: object,
  }))
}

const PlusIconContainer = styled.span`
  display: inline-block;
  vertical-align: middle;
  font-size: 1.3em;
  margin-right: 6px;
  margin-top: -3px;
`

const StyledLegendItem = styled.div`
  border-top: 2px solid #000;
  ${props =>
    props.hasSearchInterface
      ? `
    overflow: visible;
  `
      : `
    overflow-y: scroll;
  `}
  padding: 12px 15px 12px 10px;
`

const StyledLegendTitle = styled(Heading3)`
  display: block;
  margin-bottom: 20px;
`

const Dataset = styled.div`
  font-size: 1.1rem;
  margin: 15px 0;
  position: relative;
`
Dataset.displayName = 'Dataset'

const UnselectDataset = styled(DisplayText)`
  position: absolute;
  right: 0;
  width: 16px;
  color: ${v.colors.commonDark};
  transition: color 150ms;
  &:hover,
  &:active {
    color: black;
  }
`
UnselectDataset.displayName = 'UnselectDataset'

const AreaChartIcon = styled.span`
  // opacity: ${props => (props.isPrimary ? 1 : 0.8)};
  opacity: 0.8;
  display: inline-block;
  width: 100%;
  height: 100%;
  background-color: ${props => props.color};
`
AreaChartIcon.displayName = 'AreaChartIcon'

const DatasetIconWrapper = styled.span`
  display: inline-block;
  width: 15px;
  height: 15px;
  margin: 0 7px 0 5px;
`

const StyledAddComparison = styled.div`
  position: relative;
  > div {
    display: inline-block;
  }
  &:hover,
  &:active {
    color: ${v.colors.commonDarkest};
  }
`

const DatasetText = styled(SmallHelperText)`
  display: inline-block;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 240px;
  white-space: nowrap;
  @media only screen and (max-width: ${v.responsive.largeBreakpoint}px) {
    width: 200px;
  }
`

@inject('apiStore', 'uiStore')
@observer
class LegendItemCover extends React.Component {
  state = {
    comparisonMenuOpen: false,
  }

  componentDidMount() {
    this.searchTestCollections(' ')
  }

  componentWillUnMount() {
    const { uiStore } = this.props
    uiStore.removeEmptySpaceClickHandler(this.onSearchClose)
  }

  usesTestComparisonApi(entity) {
    const isTestCollection = entity.internalType === 'collections'
    // If it is a test collection, it uses the test comparison api
    if (isTestCollection) return true
    // If not a dataset, does not use test comparison API
    if (!entity.internalType === 'datasets') return false
    const hasGroupings = entity.groupings && entity.groupings.length
    // Uses test comparison API if it references a test collection,
    // and does not have any groupings
    return entity.test_collection_id && !hasGroupings
  }

  rendersAsLine(selectedDataset, isPrimary) {
    if (selectedDataset.chart_type === 'line') return true
    if (
      selectedDataset.chart_type === 'area' &&
      !isPrimary &&
      !selectedDataset.hasDates
    )
      return true
    return false
  }

  /*
   * Unique dataset and whether selected or not
   */
  datasets = ({ selected }) => {
    const { datasets } = this.props.item
    const identifiers = []
    return datasets.filter(dataset => {
      if (
        dataset.selected === selected &&
        !_.includes(identifiers, dataset.identifier)
      ) {
        identifiers.push(dataset.identifier)
        return true
      } else {
        return false
      }
    })
  }

  /*
   * Toggle's a dataset that already exists on the legend items and charts
   * to have it's data_items_dataset.selected property toggled
   */
  @observable
  @action
  toggleDatasetsWithIdentifier = async ({ identifier, selected }) => {
    const { parent } = this.props.card
    if (!selected) {
      await parent.API_selectDatasetsWithIdentifier({ identifier })
    } else {
      await parent.API_unselectDatasetsWithIdentifier({ identifier })
    }
    parent.API_fetchCards()
  }

  toggleComparisonSearch = () => {
    const { uiStore } = this.props
    const { comparisonMenuOpen } = this.state
    if (comparisonMenuOpen) {
      uiStore.removeEmptySpaceClickHandler(this.onSearchClose)
    } else {
      uiStore.addEmptySpaceClickHandler(this.onSearchClose)
    }
    this.setState({
      comparisonMenuOpen: !comparisonMenuOpen,
    })
  }

  onSearchClose = ev => {
    this.setState({
      comparisonMenuOpen: false,
    })
  }

  /*
   * When de-selecting a comparison with the Unselect button. Will always
   * pass in a dataset and will either toggle the dataset's selected property
   * if it's a grouped dataset (for now) or remove the whole data_items_dataset
   *
   * Datasets with groupings should always remain on the legend item and
   * chart items so they appear by default in the add comparisons search menu.
   * This behavior may change in the future
   */
  onDeselectComparison = async dataset => {
    const { parent } = this.props.card
    if (this.usesTestComparisonApi(dataset)) {
      await parent.API_removeComparison({ id: dataset.test_collection_id })
    } else {
      const { identifier, selected } = dataset
      this.toggleDatasetsWithIdentifier({ identifier, selected })
    }
    parent.API_fetchCards()
  }

  /*
   * When selecting a comparison with the autocomplete search. Will pass in
   * either a dataset, in the case of org-wide and test audiences, or a
   * TestCollection, in the case of the collection search.
   *
   * Datasets with groupings should always remain on the legend item and
   * chart items so they appear by default in the add comparisons search menu.
   * This behavior may change in the future
   */
  onSelectComparison = async entity => {
    const { card } = this.props
    if (this.usesTestComparisonApi(entity)) {
      await card.parent.API_addComparison(entity)
    } else {
      const { identifier, selected } = entity
      this.toggleDatasetsWithIdentifier({ identifier, selected })
    }
    card.parent.API_fetchCards()
  }

  findDatasetByTest(testId) {
    const { item } = this.props
    const dataSet = item.datasets.find(
      d => d.data_source_id === parseInt(testId)
    )
    return dataSet
  }

  searchTestCollections = (term, callback) => {
    const { item, apiStore } = this.props
    if (!apiStore.currentUser) {
      return
    }
    if (!term) {
      callback()
      return
    }
    return apiStore
      .searchCollections({
        query: `${term}`,
        type: 'Collection::TestCollection',
        order_by: 'updated_at',
        order_direction: 'desc',
        per_page: 30,
      })
      .then(res => res.data)
      .then(records => records.filter(record => record.id !== item.parent_id))
      .then(records =>
        records.filter(record => !this.findDatasetByTest(record.id))
      )
      .then(records => callback && callback(formatForAutocomplete(records)))
      .catch(e => {
        trackError(e)
      })
  }

  handleUnselectedDatasetOption = event => {
    event.preventDefault()
    const { value } = event.target
    this.toggleDatasetsWithIdentifier({ identifier: value, selected: false })
  }

  get renderDatasetsMenu() {
    return (
      <Select
        classes={{ root: 'select', selectMenu: 'selectMenu' }}
        displayEmpty
        disableUnderline
        name="role"
        onChange={this.handleUnselectedDatasetOption}
        onClose={() => this.setState({ comparisonMenuOpen: false })}
        open
        inline
      >
        {this.datasets({ selected: false }).map(dataset => (
          <SelectOption
            classes={{ root: 'selectOption', selected: 'selected' }}
            key={dataset.identifier}
            value={dataset.identifier}
          >
            {dataset.name}
          </SelectOption>
        ))}
      </Select>
    )
  }

  get renderTestCollectionsSearch() {
    // Transform the audience so name is set to display name for the option
    // formatting
    const selectedDatasetIdentifiers = this.datasets({ selected: true }).map(
      d => d.identifier
    )
    const unselectedDatasets = this.datasets({ selected: false })
    const formattedOptions = formatForAutocomplete(
      _.reject(unselectedDatasets, unselected =>
        _.includes(selectedDatasetIdentifiers, unselected.identifier)
      )
    )
    return (
      <AutoComplete
        defaultOptions={formattedOptions}
        optionSearch={this.searchTestCollections}
        onOptionSelect={option => this.onSelectComparison(option)}
        placeholder="search comparisons"
        onMenuClose={this.onSearchClose}
      />
    )
  }

  get renderComparisonMenu() {
    const { legend_search_source } = this.props.item
    if (legend_search_source === 'select_from_datasets') {
      return this.renderDatasetsMenu
    } else if (legend_search_source === 'search_test_collections') {
      return this.renderTestCollectionsSearch
    }
  }

  renderSelectedDataset = ({ dataset, order, colorOrder }) => {
    if (!dataset) return ''
    const { identifier, name, style } = dataset
    const primary = order === 0
    let icon
    if (this.rendersAsLine(dataset, primary)) {
      icon = (
        <LineChartIcon
          color={(style && style.fill) || '#000000'}
          order={order}
        />
      )
    } else {
      const { item } = this.props
      const itemStyle = item.style
      let renderedColor =
        style && style.fill
          ? color(style.fill)
              .darken(0.18 * colorOrder)
              .string()
          : colorScale[order]
      // Style on the item itself should aways override dataset style.
      if (itemStyle && itemStyle.fill) {
        renderedColor = itemStyle.fill
      }
      icon = (
        <AreaChartIcon
          color={renderedColor}
          isPrimary={primary}
          data-dataset-id={dataset.id}
        />
      )
    }
    return (
      <Dataset key={`dataset-${identifier}`}>
        {icon && <DatasetIconWrapper>{icon}</DatasetIconWrapper>}
        <DatasetText color={v.colors.black}>{name}</DatasetText>
        {!primary && dataset.selected && (
          <UnselectDataset
            role="button"
            onClick={() => this.onDeselectComparison(dataset)}
          >
            <XIcon />
          </UnselectDataset>
        )}
      </Dataset>
    )
  }

  render() {
    const { item } = this.props
    const { comparisonMenuOpen } = this.state
    let order = -1
    let colorOrder = -1
    return (
      <StyledLegendItem
        data-cy="LegendItemCover"
        hasSearchInterface={
          item.legend_search_source === 'search_test_collections'
        }
      >
        <StyledLegendTitle>{item.name}</StyledLegendTitle>
        {this.datasets({ selected: true }).map(dataset => {
          if (dataset.hasDates) colorOrder += 1
          return this.renderSelectedDataset({
            dataset,
            order: (order += 1),
            colorOrder,
          })
        })}
        <br />
        <StyledAddComparison>
          {comparisonMenuOpen && this.renderComparisonMenu}
          {item.can_edit_content && !comparisonMenuOpen && (
            <Heading3
              onClick={this.toggleComparisonSearch}
              role="button"
              className="test-add-comparison-button"
            >
              <PlusIconContainer>+</PlusIconContainer>
              Add Comparison
            </Heading3>
          )}
        </StyledAddComparison>
      </StyledLegendItem>
    )
  }
}

LegendItemCover.propTypes = {
  item: MobxPropTypes.objectOrObservableObject.isRequired,
  card: MobxPropTypes.objectOrObservableObject,
}

LegendItemCover.wrappedComponent.propTypes = {
  apiStore: MobxPropTypes.objectOrObservableObject.isRequired,
  uiStore: MobxPropTypes.objectOrObservableObject.isRequired,
}

LegendItemCover.defaultProps = {
  card: null,
}

export default LegendItemCover
