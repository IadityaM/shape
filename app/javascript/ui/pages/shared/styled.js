import { Flex } from 'reflexbox'
import styled from 'styled-components'

import v from '~/utils/variables'

export const StyledTitleAndRoles = styled(Flex)`
  .title {
    max-width: 70%;
  }
  @media only screen and (max-width: ${v.responsive.largeBreakpoint}px) {
    .title {
      max-width: 65%;
    }
  }
  @media only screen and (max-width: ${v.responsive.medBreakpoint}px) {
    .title {
      max-width: 90%;
    }
  }
  .page-menu {
    position: relative;
    top: -4px;
  }
  /* needs to be bumped down because there's no RolesSummary on items for now */
  .item-page .page-menu {
    top: 35px;
  }
  &.user-profile .title {
    cursor: pointer;
  }
`

export default {
  StyledTitleAndRoles
}
