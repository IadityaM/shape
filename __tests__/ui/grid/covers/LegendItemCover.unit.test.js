import LegendItemCover from '~/ui/grid/covers/LegendItemCover'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'

import { Heading3 } from '~/ui/global/styled/typography'
import fakeApiStore from '#/mocks/fakeApiStore'
import fakeUiStore from '#/mocks/fakeUiStore'
import {
  fakeDataset,
  fakeLegendItem,
  fakeLegendItemCard,
  fakeCollection,
} from '#/mocks/data'

let wrapper, instance, props
describe('LegendItemCover', () => {
  beforeEach(() => {
    render = () => {
      props = {
        item: fakeLegendItem,
        card: {
          ...fakeLegendItemCard,
          parent: fakeCollection,
        },
        apiStore: fakeApiStore(),
        uiStore: fakeUiStore,
      }
      wrapper = shallow(<LegendItemCover.wrappedComponent {...props} />)
      instance = wrapper.instance()
    }
    render()
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  it('shows selected dataset', () => {
    const selectedMeasure = wrapper.find('Dataset').at(0)
    expect(selectedMeasure.exists()).toBe(true)
  })

  describe('unselect', () => {
    const getUnselect = () => {
      return wrapper
        .find('Dataset')
        .at(1)
        .find('UnselectDataset')
    }

    beforeEach(() => {})

    describe('when unselecting a grouped dataset', () => {
      beforeEach(() => {
        props.item.datasets = [
          fakeDataset,
          {
            ...fakeDataset,
            groupings: [{ type: 'Organization', id: 13 }],
            selected: true,
            name: 'org-wide-question',
            display_name: 'Organziation stuff',
          },
        ]
        render()
        getUnselect().simulate('click')
      })

      it('should unselect the dataset by name', () => {
        expect(
          props.card.parent.API_unselectDatasetsWithName
        ).toHaveBeenCalled()
      })
    })

    describe('when unselecting a test dataset', () => {
      beforeEach(() => {
        props.item.datasets = [
          fakeDataset,
          { ...fakeDataset, groupings: [], selected: true, name: 'ads' },
        ]
        render()
        getUnselect().simulate('click')
      })

      it('should remove the compairison', () => {
        expect(props.card.parent.API_removeComparison).toHaveBeenCalled()
      })
    })
  })

  describe('datasets()', () => {
    beforeEach(() => {
      const datasetSelectedA = { ...fakeDataset, name: 'a', selected: true }
      const datasetSelectedB = { ...fakeDataset, name: 'b', selected: true }
      const datasetUnselectedC = { ...fakeDataset, name: 'c', selected: false }
      const datasetDuplicatedA = { ...fakeDataset, name: 'a', selected: true }
      props.item.datasets = [
        datasetSelectedA,
        datasetSelectedB,
        datasetUnselectedC,
        datasetDuplicatedA,
      ]
      render()
    })

    it('should filter by selected datasets', () => {
      expect(instance.datasets({ selected: true }).length).toEqual(2)
      expect(instance.datasets({ selected: false }).length).toEqual(1)
    })

    it('should filter out duplicates by name', () => {
      expect(instance.datasets({ selected: true }).length).toEqual(2)
    })
  })

  describe('toggle comparison search', () => {
    const comparisonClick = () => {
      wrapper.find(Heading3).simulate('click')
    }

    it('should remove/open the empty space click handler', () => {
      comparisonClick()
      expect(props.uiStore.addEmptySpaceClickHandler).toHaveBeenCalled()
    })

    it('should set the comparison menu open state', () => {
      comparisonClick()
      expect(wrapper.state().comparisonMenuOpen).toBe(true)
    })
  })

  describe('onSelectComparison', () => {
    let groupedDataset

    beforeEach(() => {
      groupedDataset = {
        ...fakeDataset,
        groupings: [{ type: 'Organization', id: 13 }],
        selected: false,
        name: 'org-wide-question',
        display_name: 'Organziation stuff',
        internalType: 'datasets',
      }

      props.item.datasets = [groupedDataset]
      render()
    })

    describe('with a test collection', () => {
      it('should add the test comparison with the test', () => {
        instance.onSelectComparison(fakeCollection)
        expect(props.card.parent.API_addComparison).toHaveBeenCalled()
      })
    })

    describe('with a dataset', () => {
      it('should toggle the dataset with name', () => {
        instance.onSelectComparison(groupedDataset)
        expect(props.card.parent.API_selectDatasetsWithName).toHaveBeenCalled()
      })
    })
  })
})
