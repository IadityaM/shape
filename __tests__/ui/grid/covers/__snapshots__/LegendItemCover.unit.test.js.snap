// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegendItemCover renders snapshot 1`] = `
<styled.div
  data-cy="LegendItemCover"
>
  <Styled(StyledHeading3)>
    My Cool Item
  </Styled(StyledHeading3)>
  <Dataset
    key="dataset-dataset"
  >
    <styled.span>
      <styled.span
        color="#c0dbde"
      />
    </styled.span>
    <Styled(SmallHelperText)
      color="#120f0e"
    />
    <UnselectDataset
      onClick={[Function]}
      role="button"
    >
      <XIcon />
    </UnselectDataset>
  </Dataset>
  <Dataset
    key="dataset-85th"
  >
    <styled.span>
      <styled.span
        color="#738091"
      />
    </styled.span>
    <Styled(SmallHelperText)
      color="#120f0e"
    />
    <UnselectDataset
      onClick={[Function]}
      role="button"
    >
      <XIcon />
    </UnselectDataset>
  </Dataset>
  <Dataset
    key="dataset-75th"
  >
    <styled.span>
      <styled.span
        color="#A1A6B4"
      />
    </styled.span>
    <Styled(SmallHelperText)
      color="#120f0e"
    />
    <UnselectDataset
      onClick={[Function]}
      role="button"
    >
      <XIcon />
    </UnselectDataset>
  </Dataset>
  <br />
  <styled.div>
    <StyledHeading3
      className="test-add-comparison-button"
      onClick={[Function]}
      role="button"
    >
      <styled.span>
        <PlusIcon
          viewBox="0 0 5 18"
        />
      </styled.span>
      Add Comparison
    </StyledHeading3>
  </styled.div>
</styled.div>
`;
